QEfficient/base/modeling_qeff.py

```diff
@@ -5,96 +5,307 @@
+            :export_kwargs (dict): Additional arguments to be passed to `torch.onnx.export`.
+            :onnx_transform_kwargs (dict): Additional arguments to be passed to `Transform.apply` for this class.
+            :export_dir (str): Specify the export directory. The export_dir will be suffixed with a hash corresponding to current model.
+        """
+        export_dir = Path(export_dir or (QEFF_HOME / self.model_name))
+        export_dir = export_dir.with_name(export_dir.name + "-" + self.model_hash)
+        onnx_path = export_dir / f"{self.model_name}.onnx"
+        if onnx_path.is_file():
+            self.onnx_path = onnx_path
+            return onnx_path
+
+        tmp_onnx_dir = export_dir / "onnx_tmp"
+        tmp_onnx_path = tmp_onnx_dir / f"{self.model_name}.onnx"
+        tmp_onnx_dir.mkdir(parents=True, exist_ok=True)
+
+        # Create input_names from example_inputs
+        input_names = []
+        for param in inspect.signature(self.model.forward).parameters:
+            if param in example_inputs:
+                if param == "past_key_values":
+                    for i in range(len(example_inputs["past_key_values"])):
+                        input_names.append(f"past_key.{i}")
+                        input_names.append(f"past_value.{i}")
+                else:
+                    input_names.append(param)
+
+        try:
+            torch.onnx.export(
+                self.model,
+                (example_inputs,),
+                str(tmp_onnx_path),
+                input_names=input_names,
+                output_names=output_names,
+                dynamic_axes=dynamic_axes,
+                opset_version=13,
+                **export_kwargs,
+            )
+            logger.info("Pytorch export successful")
+
+            model = onnx.load(tmp_onnx_path, load_external_data=False)
+            onnx_transform_kwargs = {
+                "onnx_base_dir": str(tmp_onnx_dir),
+                "model_name": self.model_name,
+                **onnx_transform_kwargs,
+            }
+            for transform in self._onnx_transforms:
+                model, transformed = transform.apply(model, **onnx_transform_kwargs)
+            model.metadata_props.append(
+                onnx.StringStringEntryProto(key="qeff_transforms", value=",".join(self._transform_names()))
+            )
+            logger.info("ONNX transforms applied")
+
+            onnx.save(model, onnx_path)
+            logger.info("Transformed onnx saved")
+
+        except Exception as e:
+            logger.error(f"ONNX export failed: {e}")
+            raise e
+
+        finally:
+            shutil.rmtree(tmp_onnx_dir, ignore_errors=True)
+
+        self.onnx_path = onnx_path
+        return onnx_path
+
+    def _compile(
+        self,
+        onnx_path: Optional[str] = None,
+        compile_dir: Optional[str] = None,
+        *,
+        specializations: Optional[List[Dict[str, int]]] = None,
```

Comment:
I wanted to make the `_compile` a core utility. Not something tied to causal LM or any other use case. So, there is no need of change this in future.

We can make the subclasses like QEffAutoModelForCausalLM have required argument, rather than enforcing in base class.